File: scraper/adapters/kayak.py
"""
scraper/adapters/kayak.py
Secondary cash-fare fallback scraper using Kayak Canada.
Only used when Google Flights returns no results.
Samples every other date to reduce load and block risk.
"""

import asyncio
import logging
import re
from datetime import datetime

from playwright.async_api import TimeoutError as PWTimeout

from scraper.monitors import CashMonitor
from scraper.adapters.base import (
    launch_browser, make_stealth_context, human_delay, retry_async
)
from scraper.db import make_fingerprint

logger = logging.getLogger(__name__)
PROVIDER = "Kayak"


def _build_url(origin: str, destination: str, date: str) -> str:
    return f"https://www.ca.kayak.com/flights/{origin}-{destination}/{date}?sort=price_a"


async def _parse_page(page, monitor: CashMonitor, date: str) -> list[dict]:
    results = []
    now = datetime.utcnow().isoformat()

    try:
        await page.wait_for_selector('[class*="nrc6"]', timeout=12000)
    except PWTimeout:
        logger.warning(f"[Kayak] No results for {monitor.id} on {date}")
        return []

    await human_delay(2000, 3000)

    try:
        cards = await page.query_selector_all('[class*="nrc6"]')
        for card in cards[:6]:
            try:
                price_el = (
                    await card.query_selector('[class*="price-text"]') or
                    await card.query_selector('[class*="mainPrice"]') or
                    await card.query_selector('[class*="Iqt3"]')
                )
                if not price_el:
                    continue
                price_text = await price_el.inner_text()
                m = re.search(r'[\d,]+', price_text)
                if not m:
                    continue
                price = float(m.group().replace(',', ''))
                if price < 30 or price > 5000:
                    continue

                airline_el = (
                    await card.query_selector('[class*="codeshares-airline-names"]') or
                    await card.query_selector('[class*="carrier-name"]') or
                    await card.query_selector('[class*="VY2U"]')
                )
                airline = (await airline_el.inner_text()).strip() if airline_el else "Unknown"

                dur_el = await card.query_selector('[class*="duration"]')
                duration = (await dur_el.inner_text()).strip() if dur_el else "N/A"

                stops_el = await card.query_selector('[class*="stops-text"]')
                stops_text = (await stops_el.inner_text()).strip().lower() if stops_el else "nonstop"
                stops = 0 if "nonstop" in stops_text else int(re.search(r'\d+', stops_text).group() if re.search(r'\d+', stops_text) else 1)

                fp = make_fingerprint(monitor.id, PROVIDER, date, airline, price, stops)
                results.append({
                    "monitor_id": monitor.id,
                    "provider": PROVIDER,
                    "kind": "cash",
                    "origin": monitor.origin,
                    "destination": monitor.destination,
                    "departure_date": date,
                    "total_price": price,
                    "currency": "CAD",
                    "airline": airline,
                    "flight_number": None,
                    "stops": stops,
                    "duration": duration,
                    "booking_url": _build_url(monitor.origin, monitor.destination, date),
                    "checked_at": now,
                    "fingerprint": fp,
                })
            except Exception as e:
                logger.debug(f"[Kayak] Card parse error: {e}")
    except Exception as e:
        logger.warning(f"[Kayak] Page eval error: {e}")

    return results


async def scrape_window(monitor: CashMonitor) -> list[dict]:
    playwright, browser = await launch_browser()
    ctx = await make_stealth_context(browser)
    all_results: list[dict] = []

    # Sample every other date to reduce footprint
    dates = monitor.dates()
    sample = dates[::2] if len(dates) > 3 else dates

    try:
        for date in sample:
            page = await ctx.new_page()
            try:
                url = _build_url(monitor.origin, monitor.destination, date)
                logger.info(f"[Kayak] {monitor.origin}→{monitor.destination} on {date}")

                async def do_scrape():
                    await page.goto(url, wait_until="domcontentloaded", timeout=30000)
                    await human_delay(2000, 4000)
                    return await _parse_page(page, monitor, date)

                results = await retry_async(do_scrape, retries=2, base_delay=7.0)
                all_results.extend(results)
                logger.info(f"[Kayak] → {len(results)} results")
            except Exception as e:
                logger.warning(f"[Kayak] Failed on {date}: {e}")
            finally:
                await page.close()

            await human_delay(4000, 7000)

    finally:
        await ctx.close()
        await browser.close()
        await playwright.stop()

    return all_results

File: apps/worker/src/adapters/kayak.ts
// apps/worker/src/adapters/kayak.ts
import { chromium, Browser, BrowserContext } from "playwright";
import { CashMonitor, AwardMonitor, FlightResult } from "@flight-tracker/shared";
import { ScrapeAdapter, buildFingerprint, sleep, retryWithBackoff } from "./base";

const PROXY_CONFIG = process.env.PROXY_ENDPOINT
  ? {
      server: process.env.PROXY_ENDPOINT,
      username: process.env.PROXY_USERNAME,
      password: process.env.PROXY_PASSWORD,
    }
  : undefined;

export class KayakAdapter implements ScrapeAdapter {
  name = "Kayak";
  priority = 3;

  private browser: Browser | null = null;
  private context: BrowserContext | null = null;

  async initialize(): Promise<void> {
    this.browser = await chromium.launch({
      headless: true,
      args: ["--no-sandbox", "--disable-setuid-sandbox"],
    });
    this.context = await this.browser.newContext({
      proxy: PROXY_CONFIG,
      userAgent:
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36",
      viewport: { width: 1280, height: 800 },
      locale: "en-CA",
    });
    await this.context.addInitScript(() => {
      Object.defineProperty(navigator, "webdriver", { get: () => false });
    });
  }

  async isHealthy(): Promise<boolean> {
    return true;
  }

  async scrapeWindow(monitor: CashMonitor): Promise<FlightResult[]> {
    if (!this.browser) await this.initialize();
    const results: FlightResult[] = [];

    const dates = this.getDateRange(monitor.dateFrom, monitor.dateTo);
    // Sample a few dates from the window to avoid excessive requests
    const sampleDates = dates.filter((_, i) => i % 2 === 0);

    for (const date of sampleDates) {
      try {
        const dayResults = await retryWithBackoff(
          () => this.scrapeDate(monitor, date),
          2,
          6000
        );
        results.push(...dayResults);
        await sleep(4000 + Math.random() * 3000);
      } catch (err: any) {
        console.warn(`[Kayak] ${monitor.id} on ${date}: ${err.message}`);
      }
    }
    return results;
  }

  private async scrapeDate(
    monitor: CashMonitor,
    date: string
  ): Promise<FlightResult[]> {
    const page = await this.context!.newPage();
    const results: FlightResult[] = [];

    try {
      const url = `https://www.ca.kayak.com/flights/${monitor.origin}-${monitor.destination}/${date}?sort=price_a`;
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 30000 });
      await sleep(3000 + Math.random() * 2000);

      try {
        await page.waitForSelector('[class*="nrc6"]', { timeout: 10000 });
      } catch {
        return [];
      }

      await sleep(2000);

      const flights = await page.evaluate(() => {
        const cards = document.querySelectorAll('[class*="nrc6"]');
        const extracted: any[] = [];
        cards.forEach((card) => {
          try {
            const priceEl = card.querySelector('[class*="price-text"]') ||
              card.querySelector('[class*="mainPrice"]');
            const priceText = priceEl?.textContent ?? "";
            const priceMatch = priceText.match(/[\d,]+/);
            if (!priceMatch) return;
            const price = parseInt(priceMatch[0].replace(",", ""), 10);

            const airlineEl = card.querySelector('[class*="codeshares-airline-names"]') ||
              card.querySelector('[class*="carrier-name"]');
            const airline = airlineEl?.textContent?.trim() ?? "Unknown";

            const durationEl = card.querySelector('[class*="duration"]');
            const duration = durationEl?.textContent?.trim() ?? "Unknown";

            const stopsEl = card.querySelector('[class*="stops-text"]');
            const stopsText = stopsEl?.textContent ?? "nonstop";
            const stops = stopsText.toLowerCase().includes("nonstop") ? 0 :
              parseInt(stopsText.match(/\d+/)?.[0] ?? "1", 10);

            extracted.push({ price, airline, duration, stops });
          } catch {}
        });
        return extracted;
      });

      for (const f of flights.slice(0, 5)) {
        const result: FlightResult = {
          provider: this.name,
          monitorId: monitor.id,
          kind: "cash",
          origin: monitor.origin,
          destination: monitor.destination,
          departureDate: date,
          totalPrice: f.price,
          currency: "CAD",
          airline: f.airline,
          stops: f.stops,
          duration: f.duration,
          bookingUrl: url,
          scrapedAt: new Date().toISOString(),
        };
        (result as any).fingerprint = buildFingerprint(result);
        results.push(result);
      }
    } finally {
      await page.close();
    }

    return results;
  }

  async scrapeMonth(_monitor: AwardMonitor): Promise<FlightResult[]> {
    return [];
  }

  async close(): Promise<void> {
    await this.context?.close();
    await this.browser?.close();
    this.browser = null;
    this.context = null;
  }

  private getDateRange(from: string, to: string): string[] {
    const dates: string[] = [];
    const cur = new Date(from);
    const end = new Date(to);
    while (cur <= end) {
      dates.push(cur.toISOString().split("T")[0]);
      cur.setDate(cur.getDate() + 1);
    }
    return dates;
  }
}
